File: task_runner_service/frontend/src/types/memory.ts
export interface EngineThread {
  id: string;
  tenant_id: string;
  source_id: string;
  subject_id: string;
  thread_type: string;
  external_thread_id?: string | null;
  title?: string | null;
  labels?: string[] | null;
  metadata?: unknown;
  status: string;
  summary_status: string;
  summary_job_run_id?: string | null;
  summary_locked_at?: string | null;
  summary_lock_expires_at?: string | null;
  pending_record_count: number;
  pending_summary_tokens: number;
  created_at: string;
  updated_at: string;
  archived_at?: string | null;
}

export interface EngineRecord {
  id: string;
  thread_id: string;
  tenant_id: string;
  source_id: string;
  external_record_id?: string | null;
  role: string;
  record_type: string;
  content: string;
  structured_payload?: unknown;
  metadata?: unknown;
  summary_status: string;
  summary_id?: string | null;
  summarized_at?: string | null;
  created_at: string;
}

export interface ComposeContextBlock {
  block_type: string;
  text: string;
}

export interface ComposeContextMeta {
  summary_count: number;
  recent_record_count: number;
}

export interface ComposeContextResponse {
  thread_id: string;
  blocks: ComposeContextBlock[];
  recent_records: EngineRecord[];
  meta: ComposeContextMeta;
}

export interface TaskMemoryContextResponse {
  task_id: string;
  memory_thread_id: string;
  tenant_id: string;
  subject_id: string;
  thread?: EngineThread | null;
  context?: ComposeContextResponse | null;
  total_record_count: number;
}

export interface TaskMemoryRecordsResponse {
  task_id: string;
  memory_thread_id: string;
  tenant_id: string;
  subject_id: string;
  thread?: EngineThread | null;
  total: number;
  limit: number;
  offset: number;
  order: string;
  role?: string | null;
  record_type?: string | null;
  summary_status?: string | null;
  has_more: boolean;
  items: EngineRecord[];
}

export interface TaskMemorySummaryJobResult {
  thread_id: string;
  accepted: boolean;
  running: boolean;
  job_run_id?: string | null;
  generated: boolean;
  summary_id?: string | null;
  source_record_count: number;
}

export interface TaskMemorySummaryResponse {
  task_id: string;
  memory_thread_id: string;
  tenant_id: string;
  requested_at: string;
  result: TaskMemorySummaryJobResult;
}

File: task_runner_service/backend/src/models/memory.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// Required Notice: Copyright (c) 2025 AI Chat Team

use memory_engine_sdk::{
    ComposeContextResponse, EngineRecord, EngineThread, RunThreadRepairSummaryResponse,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TaskMemoryContextOptions {
    pub include_recent_records: Option<bool>,
    pub include_thread_summary: Option<bool>,
    pub include_subject_memory: Option<bool>,
    pub recent_record_limit: Option<usize>,
    pub summary_limit: Option<usize>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TaskMemoryRecordsOptions {
    pub role: Option<String>,
    pub record_type: Option<String>,
    pub summary_status: Option<String>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
    pub order: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskMemoryContextResponse {
    pub task_id: String,
    pub memory_thread_id: String,
    pub tenant_id: String,
    pub subject_id: String,
    pub thread: Option<EngineThread>,
    pub context: Option<ComposeContextResponse>,
    pub total_record_count: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskMemoryRecordsResponse {
    pub task_id: String,
    pub memory_thread_id: String,
    pub tenant_id: String,
    pub subject_id: String,
    pub thread: Option<EngineThread>,
    pub total: i64,
    pub limit: i64,
    pub offset: i64,
    pub order: String,
    pub role: Option<String>,
    pub record_type: Option<String>,
    pub summary_status: Option<String>,
    pub has_more: bool,
    pub items: Vec<EngineRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskMemorySummaryResponse {
    pub task_id: String,
    pub memory_thread_id: String,
    pub tenant_id: String,
    pub requested_at: String,
    pub result: RunThreadRepairSummaryResponse,
}
